source/modulo_core/include/modulo_core/communication/MessagePair.hpp

```diff
@@ -62,37 +115,67 @@ inline MsgT MessagePair<MsgT, DataT>::write_message() const {
   if (this->data_ == nullptr) {
     throw exceptions::NullPointerException("The message pair data is not set, nothing to write");
   }
+
+  MsgT message;
+  if constexpr (CustomT<MsgT> && CustomT<DataT>) {
+    message = write_raw_message();
+  } else if constexpr (std::same_as<MsgT, EncodedState>) {
+    message = write_encoded_message();
+  } else {
+    message = write_translated_message();
+  }
+  return message;
+}
+
+template<typename MsgT, typename DataT>
+inline MsgT MessagePair<MsgT, DataT>::write_translated_message() const {
   auto message = MsgT();
   translators::write_message(message, *this->data_, clock_->now());
   return message;
 }
 
 template<>
-inline EncodedState MessagePair<EncodedState, state_representation::State>::write_message() const {
-  if (this->data_ == nullptr) {
```

Comment:
Sure about that? I missed that so far

Reply:
you mean removing the nullptr check or the entire function?

Reply:
removing the nullptr check

Reply:
sorry its somewhere else now, i misread the diff

Reply:
yup, not a good idea, I thought it's a duplicate check actually

Reply:
ah, but it is. We check in write_message, before calling write_encoded_message (line 115)